src/python_ml_project_template/models/flow_trajectory_predictor.py

```diff
@@ -0,0 +1,227 @@
+import torch_geometric.data as tgd
+from flowbot3d.grasping.agents.flowbot3d import FlowNetAnimation
+
+# from flowbot3d.models.artflownet import artflownet_loss, flow_metrics
+from flowbot3d.models.artflownet import artflownet_loss
+from plotly.subplots import make_subplots
+from torch import optim
+
+
+def flow_metrics(pred_flow, gt_flow):
+    with torch.no_grad():
+        # RMSE
+        rmse = (pred_flow - gt_flow).norm(p=2, dim=-1).mean()
+
+        # Cosine similarity, normalized.
+        nonzero_gt_flowixs = torch.where(gt_flow.norm(dim=-1) != 0.0)
+        gt_flow_nz = gt_flow[nonzero_gt_flowixs]
+        pred_flow_nz = pred_flow[nonzero_gt_flowixs]
+        cos_dist = torch.cosine_similarity(pred_flow_nz, gt_flow_nz, dim=-1).mean()
+
+        # Magnitude
+        mag_error = (
+            (pred_flow.norm(p=2, dim=-1) - gt_flow.norm(p=2, dim=-1)).abs().mean()
+        )
+    return rmse, cos_dist, mag_error
+
+
+def make_trajectory_animation(traj_data):  # Make trajectory animation
+    animation = FlowNetAnimation()
+    for i in range(traj_data["point"].shape[-2]):
+        pcd = (
+            traj_data["pos"].detach().numpy()
+            if i == 0
+            else traj_data["point"][:, (i - 1)].detach().numpy()
+        )
+        mask = traj_data["mask"].detach().numpy()
+        flow = traj_data["delta"][:, i].detach().numpy()
+        animation.add_trace(
+            torch.as_tensor(pcd),
+            torch.as_tensor([pcd]),
+            torch.as_tensor([flow]),
+            "red",
+        )
+
+    return animation.animate()
+
+
+# Flow predictor
+class FlowTrajectoryTrainingModule(L.LightningModule):
+    def __init__(self, network, training_cfg) -> None:
+        super().__init__()
+        self.network = network
+        self.lr = training_cfg.lr
+        self.batch_size = training_cfg.batch_size
+        self.mask_input_channel = training_cfg.mask_input_channel
+        self.mode = training_cfg.mode
+        self.trajectory_len = training_cfg.trajectory_len
+        assert self.mode in ["delta", "point"]
+
+    def forward(self, data) -> torch.Tensor:  # type: ignore
+        # Maybe add the mask as an input to the network.
+        if self.mask_input_channel:
+            data.x = data.mask.reshape(len(data.mask), 1)
+
+        # Run the model.
+        flow = typing.cast(torch.Tensor, self.network(data))
+
+        return flow
+
+    def _step(self, batch: tgd.Batch, mode):
+        # Make a prediction.
+        f_pred = self(batch)
+
+        # Compute the loss.
+        n_nodes = torch.as_tensor([d.num_nodes for d in batch.to_data_list()]).to(self.device)  # type: ignore
+        f_ix = batch.mask.bool()
+        if self.mode == "delta":
+            f_target = batch.delta.reshape(batch.delta.shape[0], -1)
+        elif self.mode == "point":
+            f_target = batch.point.reshape(batch.point.shape[0], -1)
+
+        f_target = f_target.float()
+        loss = artflownet_loss(f_pred, f_target, n_nodes)
```

Comment:
reminder to change the loss here to be:

$$\mathcal{L} = \sum_{k=1}^K{||\hat{f_k} - f_k}||_2$$

instead of 

$$\mathcal{L} = ||\hat{f_{1...k}} - {f_{1...k}}||_2$$